XFree86-3.3/xc/programs/Xserver/hw/xfree86/doc/man/xf86dga.man
.\" Copyright (c) 1996  The XFree86 Project
.\" $XFree86: xc/programs/Xserver/hw/xfree86/doc/man/XF86DGA.man,v 3.5.2.1 1997/05/21 15:02:36 dawes Exp $
.\" 
.de ZN
.ie t \fB\^\\$1\^\fR\\$2
.el \fI\^\\$1\^\fP\\$2
..
.TH XF86DGA 3X11 "3.3" "XFree86" "X FUNCTIONS"
.SH NAME
XF86DGAQueryExtension, XF86DGAQueryVersion, XF86DGAQueryDirectVideo, XF86DGAGetVideo, XF86DGADirectVideo, XF86DGASetVidPage, XF86DGASetViewPort, XF86DGAViewPortChanged, XF86DGAGetViewPortSize, XF86DGAInstallColormap, XF86DGAForkApp \- XFree86-DGA extension interface functions
.SH SYNTAX
.nf
.LP
#include <X11/extensions/xf86dga.h>
.LP
Bool XF86DGAQueryExtension(
    Display *\fIdisplay\fP\^,
    int *\fIevent_base_return\fP\^,
    int *\fIerror_base_return\fP\^);
.LP
Bool XF86DGAQueryVersion(
    Display *\fIdisplay\fP\^,
    int *\fImajor_version_return\fP\^,
    int *\fIminor_version_return\fP\^);
.LP
Bool XF86DGAQueryDirectVideo(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int *\fIflags_return\fP\^);
.LP
Bool XF86DGAGetVideo(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    char **\fIaddr_return\fP\^,
    int *\fIwidth_return\fP\^,
    int *\fIbanksize_return\fP\^,
    int *\fImemsize_return\fP\^);
.LP
Bool XF86DGADirectVideo(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int \fIflags\fP\^);
.LP
Bool XF86DGASetVidPage(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int \fIpage\fP\^);
.LP
Bool XF86DGASetViewPort(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int \fIx\fP\^,
    int \fIy\fP\^);
.LP
Bool XF86DGAViewPortChanged(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int \fInum_pages\fP\^);
.LP
Bool XF86DGAGetViewPortSize(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    int *\fIviewport_width_return\fP\^,
    int *\fIviewport_height_return\fP\^);
.LP
Bool XF86DGAInstallColormap(
    Display *\fIdisplay\fP\^,
    int \fIscreen\fP\^,
    Colormap \fIcmap\fP\^);
.LP
int XF86DGAForkApp(
    int \fIscreen\fP\^);
.fi
.SH ARGUMENTS
.IP \fIdisplay\fP 2i
Specifies the connection to the X server.
.IP \fIscreen\fP 2i
Specifies which screen number the setting apply to.
.IP \fIevent_base_return\fP 2i
Returns the base event number for the extension.
.IP \fIerror_base_return\fP 2i
Returns the base error number for the extension.
.IP \fImajor_version_return\fP 2i
Returns the major version number of the extension.
.IP \fIminor_version_return\fP 2i
Returns the minor version number of the extension.
.IP \fIaddr_return\fP 2i
Returns a pointer to the start of the video framebuffer.
.IP \fIwidth_return\fP 2i
Returns the framebuffer line width.
.IP \fIbanksize_return\fP 2i
Returns the framebuffer bank size.
.IP \fImemsize_return\fP 2i
Returns the size of the framebuffer memory.
.IP \fIflags\fP 2i
Sets the DirectVideo access features.  When zero, DirectVideo mode is
disabled.  \fIflags\fP may be a a bit-wise combination of the following
values:
.nf
XF86DGADirectGraphics          enable Direct Video mode
XF86DGADirectMouse             enable reporting of pointer movement
                               as relative motion
XF86DGADirectKeyb              enable direct reporting of keyboard events
.fi
.IP \fIflags_return\fP 2i
Reports the DirectVideo features supported by the hardware.  When zero,
the hardware does not support direct video at all.
\fIflags\fP may be a a bit-wise combination of the following
values:
.nf
XF86DGADirectPresent           DirectVideo support is present
.fi
.IP \fIpage\fP 2i
Indicates the framebuffer page (bank) to activate for read/write access.
.IP \fIx\fP 2i
Incidates the x coordinate for the upper-left corner of the view port.
.IP \fIy\fP 2i
Incidates the y coordinate for the upper-left corner of the view port.
.IP \fIviewport_width_return\fP 2i
Reports the width of the view port.
.IP \fIviewport_width_return\fP 2i
Reports the height of the view port.
.IP \fInum_pages\fP 2i
Indicates the number of pages when doing hardware multi-buffering.
.IP \fIcmap\fP 2i
Indicates the colormap to install.
.SH DESCRIPTION
These functions provide an interface to the server extension
\fIXFree86-DGA\fP
which allows a local client direct access to the video framebuffer.
Applications that use these functions must be linked with
.ZN -lXxf86dga
.SS "DGA FUNCTIONS"
The function
.ZN XF86DGAQueryDirectVideo
returns the DirectVideo capabilities supported by the graphics device.
.PP
The
.ZN XF86DGAGetVideo
function is used to get the parameters for the framebuffer.  The parameters
returned are a pointer to the start of the framebuffer, the framebuffer line
width, framebuffer bank size, and framebuffer memory size.
.PP
.ZN XF86DGADirectVideo
is used to enable or disable DirectVideo mode.
.PP
The
.ZN XF86DGASetVidPage
function sets the currently active framebuffer page (bank).  This is only
required for hardware which has a banked memory layout (banksize < memsize).
.PP
The
.ZN XF86DGASetViewPort
function sets the framebuffer coordinates to use for the upper-left corner
of the view port.
.PP
The
.ZN XF86DGAViewPortChanged
function checks whether a previous XF86DGASetViewPort command has been
completed by the hardware, that is, whether a vertical retrace has
occurred since a previous XF86DGASetViewPort.  This can (must, in fact)
be used with page-flipping; you can start writing to the next page only
when this function returns TRUE.  For some devices this will be the case
immediately after XF86DGASetViewPort, however this may be changed in the
future.  The number of pages used is specified with \fInum_pages\fP; it
should be 2 for simple page flipping (double buffering).  If n is greater
than two (triple or multi-buffering), the function checks whether the
(n \- 2)-before-last SetViewPort has been completed.
.PP
The
.ZN XF86DGAGetViewPortSize
function returns the size of the view port, which is the part of the
framebuffer that is visible on the screen.
.PP
The
.ZN XF86DGAInstallColormap
function is used to install a colormap.  This must be called after DirectVideo
mode has been enabled.
.PP
The
.ZN XF86DGAForkApp
function causes the client application to fork, leaving the parent process
to hang around and return to non-DGA mode should the client exit for any
reason.  This function returns 0 for success, or the error condition
returned by fork().
.SS "OTHER FUNCTIONS"
The
.ZN XF86DGAQueryVersion
function can be used to determine the version of the extension
built into the server.
.PP
The function
.ZN XF86DGAQueryExtension
returns the lowest numbered error and event values
assigned to the extension.
.SH SEE ALSO
XFree86(1), XF86Config(4/5)
.SH AUTHORS
Jon Tombs, Harm Hanemaayer, Mark Vojkovich.



ace/ChangeLog
Fri Oct 05 05:56:50 2001  Balachandran  <bala@cs.wustl.edu>

        * ACE version 5.2 released.

Thu Oct  4 17:38:21 2001  Steve Huston  <shuston@riverace.com>

        * ace/Synch.h: Added some comments to explain the rationale for
          the different return values from the blocking and timed
          versions of ACE_Null_Semaphore::acquire().  Thanks to Doug
          Schmidt for explaining this.

Thu Oct  4 17:27:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp (recv_log_record): Insure that
          the payload message block is properly aligned for any CDR
          extraction. Many thanks to Bala for noticing this problem and
          sending code for the fix.

Thu Oct  4 17:08:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Process_Per_Connection_Logging_Server.cpp:
          In prepare(), call ACE_Process_Options::command_line() rather
          than ACE_Process_Options::process_name() to set the program
          path for the child process.

Thu Oct  4 13:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp: Use a ACE_Message_Block,
          rather than hijacking a CDR block, to read the payload into.

Thu Oct  4 10:14:34 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Donald Acton to the Hall of Fame.

Thu Oct  4 06:39:20 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/config-sunos5.5.h: Added ACE_LACKS_MKSTEMP as SunOS 5.5 does
	  not support mkstemp (). 
	
	* ace/config-sunos5.6.h: Undefined ACE_LACKS_MKSTEMP as Sun 5.6
	  and above support it. 

	  Thanks to Don Hinton <dhinton@swan.fr> for reporting this
	  problem and Chad Elliott for testing this out on a Sun OS 5.5
	  box. 

Wed Oct  3 16:37:41 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (read_1): Reverted this change "Wed Oct  3
	  15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>".  This
	  change is alone not enough to fix the problem that Brad
	  reported. The sample program for the text needs to be fixed to
	  get it working right.

Wed Oct  3 15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>

        * ace/CDR_Stream.cpp (read_1);

          Reverted changed made in 

          Fri Aug 17 16:56:16 2001  Jeff Parsons <parsons@cs.wustl.edu>

          to read_1(). A similar change made in several other methods
          is ok, but in this case bytes can be copied into the
          input CDR stream without changing wr_ptr, so the check will
          fail unnecessarily. Thanks to Brad Hoskins <BHoskins@slo.newport.com>
          for pointing this out.

Wed Oct  2 12:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Server.cpp: Fixed spelling error.
        * examples/C++NPv1/Logging_Handler.cpp: Comment length adjustment.

Tue Oct  2 21:52:31 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Reverted this:
          Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

Tue Oct  2 19:52:57 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* Makefile: Removed some of the files that no longer exist in the
	  repository. The inclusion of these files in the Makefile added
	  noise during the beta cutting process.

Tue Oct 02 15:34:49 2001  Carlos O'Ryan  <coryan@uci.edu>

	* etc/tao_cosevent.doxygen:
	* etc/tao_cosnotification.doxygen:
	  Fixed typos in the tag file names.

Tue Oct  2 15:52:27 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Sunny Leung to the Hall of Fame.

Tue Oct  2 15:33:36 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Removed one instance of Peter Phillips
	  <pphillip@opentext.com> from the Hall of Fame.

Tue Oct  2 15:00:37 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (clone_from): Copied the GIOP major and minor
	  versions from the incoming CDR stream. Thanks to Sunny Leung 
	  <sleung@opentext.com> for reporting this. This closes bugzilla
	  #1043.  

Tue Oct  2 09:45:39 2001  Steve Huston  <shuston@riverace.com>

        * tests/libDLL_Test.icc:
        * tests/libService_Config_DLL.icc: Add link(exportAll) to the link
          options to export the entrypoints needed for the tests.

Mon Oct  1 20:04:14 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* THANKS: 
	  Added Tushar Nair to the THANKS file.

Mon Oct  1 18:06:48 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* ace/Object_Manager.cpp (ACE_UnhandledExceptionFilter):
	  Removed the un-needed #if 0's. Left them before as a safety
	  measure (in case something doesnt work) and forgot to remove
	  them later.

	* ace/Log_Msg.cpp (instance):
	  Fixed the seg faults that are raised when ace is compiled with
	  ACE_NTRACE=0 on SunOS. Added a check before locking and
	  releasing a lock.

	  Thanks to Tushar Nair <ntushar@yahoo.com> for reporting the
	  problem and for providing a fix.

Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Remove the third (optional)
          arg from getline(). Using a char literal without further
          qualification causes confusion in Unicode builds.

Mon Oct  1 14:44:13 2001  Ossama Othman  <ossama@uci.edu>

	* ace/SSL/SSL_Context.cpp (ssl_library_init):

	  Initialize the locking callbacks in the crypto library before
	  initializing the OpenSSL library.  OpenSSL may require the
	  locking callbacks to be initialized before its own
	  initialization.

	* ace/SSL/SSL_Context.cpp (ssl_library_fini):

	  Perform crypto library resource cleanup before reclaiming the
	  locks used by the locking callbacks.  Crypto library resource
	  cleanup makes use of the locking callbacks.

Mon Oct  1 15:19:11 2001  Douglas C. Schmidt  <schmidt@macarena.cs.wustl.edu>


XFree86-3.3/xc/lib/Xxf86dga/xf86dga.c
/* $XFree86: xc/lib/Xxf86dga/XF86DGA.c,v 3.11 1997/01/18 06:52:20 dawes Exp $ */
/*

Copyright (c) 1995  Jon Tombs
Copyright (c) 1995,1996  The XFree86 Project, Inc

*/

/* THIS IS NOT AN X CONSORTIUM STANDARD */

#ifdef __EMX__ /* needed here to override certain constants in X headers */
#define INCL_DOS
#define INCL_DOSIOCTL
#include <os2.h>
#endif

#define NEED_EVENTS
#define NEED_REPLIES
#include "Xlibint.h"
#include "xf86dgastr.h"
#include "Xext.h"
#include "extutil.h"

static XExtensionInfo _xf86dga_info_data;
static XExtensionInfo *xf86dga_info = &_xf86dga_info_data;
static char *xf86dga_extension_name = XF86DGANAME;

#define XF86DGACheckExtension(dpy,i,val) \
  XextCheckExtension (dpy, i, xf86dga_extension_name, val)

/*****************************************************************************
 *                                                                           *
 *			   private utility routines                          *
 *                                                                           *
 *****************************************************************************/

static int close_display();
static /* const */ XExtensionHooks xf86dga_extension_hooks = {
    NULL,				/* create_gc */
    NULL,				/* copy_gc */
    NULL,				/* flush_gc */
    NULL,				/* free_gc */
    NULL,				/* create_font */
    NULL,				/* free_font */
    close_display,			/* close_display */
    NULL,				/* wire_to_event */
    NULL,				/* event_to_wire */
    NULL,				/* error */
    NULL,				/* error_string */
};

static XEXT_GENERATE_FIND_DISPLAY (find_display, xf86dga_info, 
				   xf86dga_extension_name, 
				   &xf86dga_extension_hooks, 
				   0, NULL)

static XEXT_GENERATE_CLOSE_DISPLAY (close_display, xf86dga_info)


/*****************************************************************************
 *                                                                           *
 *		    public XFree86-DGA Extension routines                *
 *                                                                           *
 *****************************************************************************/

Bool XF86DGAQueryExtension (dpy, event_basep, error_basep)
    Display *dpy;
    int *event_basep, *error_basep;
{
    XExtDisplayInfo *info = find_display (dpy);

    if (XextHasExtension(info)) {
	*event_basep = info->codes->first_event;
	*error_basep = info->codes->first_error;
	return True;
    } else {
	return False;
    }
}

Bool XF86DGAQueryVersion(dpy, majorVersion, minorVersion)
    Display* dpy;
    int* majorVersion; 
    int* minorVersion;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAQueryVersionReply rep;
    xXF86DGAQueryVersionReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAQueryVersion, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAQueryVersion;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }
    *majorVersion = rep.majorVersion;
    *minorVersion = rep.minorVersion;
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}

Bool XF86DGAGetVideoLL(dpy, screen, offset, width, bank_size, ram_size)
    Display* dpy;
    int screen;
    int *offset;
    int *width, *bank_size, *ram_size;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAGetVideoLLReply rep;
    xXF86DGAGetVideoLLReq *req;
    int i;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAGetVideoLL, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAGetVideoLL;
    req->screen = screen;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }

    *offset = /*(char *)*/rep.offset;
    *width = rep.width;
    *bank_size = rep.bank_size;
    *ram_size = rep.ram_size;
	
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}

    
Bool XF86DGADirectVideoLL(dpy, screen, enable)
    Display* dpy;
    int screen;
    int enable;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGADirectVideoReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGADirectVideo, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGADirectVideo;
    req->screen = screen;
    req->enable = enable;
    UnlockDisplay(dpy);
    SyncHandle();
    XSync(dpy,False);
    return True;
}

Bool XF86DGAGetViewPortSize(dpy, screen, width, height)
    Display* dpy;
    int screen;
    int *width, *height;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAGetViewPortSizeReply rep;
    xXF86DGAGetViewPortSizeReq *req;
    int i;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAGetViewPortSize, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAGetViewPortSize;
    req->screen = screen;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }

    *width = rep.width;
    *height = rep.height;
	
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}
    
    
Bool XF86DGASetViewPort(dpy, screen, x, y)
    Display* dpy;
    int screen;
    int x, y;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGASetViewPortReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGASetViewPort, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGASetViewPort;
    req->screen = screen;
    req->x = x;
    req->y = y;
    UnlockDisplay(dpy);
    SyncHandle();
    XSync(dpy,False);
    return True;
}

    
Bool XF86DGAGetVidPage(dpy, screen, vpage)
    Display* dpy;
    int screen;
    int *vpage;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAGetVidPageReply rep;
    xXF86DGAGetVidPageReq *req;
    int i;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAGetVidPage, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAGetVidPage;
    req->screen = screen;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }

    *vpage = rep.vpage;
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}

    
Bool XF86DGASetVidPage(dpy, screen, vpage)
    Display* dpy;
    int screen;
    int vpage;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGASetVidPageReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGASetVidPage, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGASetVidPage;
    req->screen = screen;
    req->vpage = vpage;
    UnlockDisplay(dpy);
    SyncHandle();
    XSync(dpy,False);
    return True;
}

Bool XF86DGAInstallColormap(dpy, screen, cmap)
Display* dpy;
int screen;
Colormap cmap;
{
   XExtDisplayInfo *info = find_display (dpy);
   xXF86DGAInstallColormapReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAInstallColormap, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAInstallColormap;
    req->screen = screen;
    req->id = cmap;
    UnlockDisplay(dpy);
    SyncHandle();
    XSync(dpy,False);
    return True;
}

Bool XF86DGAQueryDirectVideo(dpy, screen, flags)
    Display *dpy;
    int screen;
    int *flags;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAQueryDirectVideoReply rep;
    xXF86DGAQueryDirectVideoReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAQueryDirectVideo, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAQueryDirectVideo;
    req->screen = screen;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }
    *flags = rep.flags;
    UnlockDisplay(dpy);
    SyncHandle();
    return True;
}

Bool XF86DGAViewPortChanged(dpy, screen, n)
    Display *dpy;
    int screen;
    int n;
{
    XExtDisplayInfo *info = find_display (dpy);
    xXF86DGAViewPortChangedReply rep;
    xXF86DGAViewPortChangedReq *req;

    XF86DGACheckExtension (dpy, info, False);

    LockDisplay(dpy);
    GetReq(XF86DGAViewPortChanged, req);
    req->reqType = info->codes->major_opcode;
    req->dgaReqType = X_XF86DGAViewPortChanged;
    req->screen = screen;
    req->n = n;
    if (!_XReply(dpy, (xReply *)&rep, 0, xFalse)) {
	UnlockDisplay(dpy);
	SyncHandle();
	return False;
    }
    UnlockDisplay(dpy);
    SyncHandle();
    return rep.result;
}



/* Helper functions */

#if 0
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#include <X11/StringDefs.h>
#include <X11/Xatom.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Scrollbar.h>
#include <X11/Xaw/Label.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xaw/Box.h>
#include <X11/Xaw/Toggle.h>
#include <X11/Xmu/StdSel.h>
#endif
#include <X11/Xmd.h>
#include <X11/extensions/xf86dga.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#if defined(ISC) 
# define HAS_SVR3_MMAP
# include <sys/types.h>
# include <errno.h>

# include <sys/at_ansi.h>
# include <sys/kd.h>

# include <sys/sysmacros.h>
# include <sys/immu.h>
# include <sys/region.h>

# include <sys/mmap.h>
#else
# if !defined(Lynx)
#  if !defined(__EMX__)
#   include <sys/mman.h>
#  endif
# else
#  include <sys/types.h>
#  include <errno.h>
#  include <smem.h>
# endif
#endif
#include <sys/wait.h>
#include <signal.h>
extern int errno;

#if defined(SVR4) && !defined(sun) && !defined(SCO325)
#define DEV_MEM "/dev/pmem"
#else
#define DEV_MEM "/dev/mem"
#endif

#if defined(ISC) && defined(HAS_SVR3_MMAP)
struct kd_memloc XFree86mloc;
#endif

static char * _XFree86addr = NULL;
static int    _XFree86size = 0;

/*
 * Still need to find a clean way of detecting the death of a DGA app
 * and returning things to normal - Jon
 * This is here to help debugging without rebooting... Also C-A-BS
 * should restore text mode.
 */

int XF86DGAForkApp(int screen)
{
     pid_t pid;
     int status;

     /* fork the app, parent hangs around to clean up */
     if ((pid = fork()) > 0) {
        Display *disp;

	waitpid(pid, &status, 0);
	disp = XOpenDisplay(NULL);
	XF86DGADirectVideo(disp, screen, 0);
	XSync(disp,False);
        if (WIFEXITED(status))
	    _exit(0);
	else
	    _exit(-1);
     }
     return pid;
}

XF86DGADirectVideo(dis, screen, enable)
Display *dis;
int screen;
int enable;
{
   if (enable&XF86DGADirectGraphics) {
	 fprintf(stderr, "video memory unprotecting\n");
#if !defined(ISC) && !defined(HAS_SVR3_MMAP) && !defined(Lynx) && !defined(__EMX__)
      if (_XFree86addr && _XFree86size)
         if (mprotect(_XFree86addr,_XFree86size, PROT_READ|PROT_WRITE)) {
         fprintf(stderr, "XF86DGADirectVideo: mprotect (%s)\n",
                           strerror(errno));
		exit (-3);
	 }
#endif
   } else {
      if (_XFree86addr && _XFree86size)
	 fprintf(stderr, "video memory protecting\n");
#if !defined(ISC) && !defined(HAS_SVR3_MMAP) && !defined(__EMX__)
#ifndef Lynx
         if (mprotect(_XFree86addr,_XFree86size, PROT_READ)) {
         fprintf(stderr, "XF86DGADirectVideo: mprotect (%s)\n",
                           strerror(errno));
		exit (-4);
	 }
#else
	smem_create(NULL, _XFree86addr, _XFree86size, SM_DETACH);
	smem_remove("XF86DGA");
#endif
#endif
   }
   XF86DGADirectVideoLL(dis, screen, enable);
}


static void XF86cleanup(int sig)
{
        Display *disp;
	static beenhere = 0;

	if (beenhere)
		_exit(3);
	beenhere = 1;
	disp = XOpenDisplay(NULL);
	XF86DGADirectVideo(disp, 0, 0);
	XSync(disp,False);
        _exit(3);
}

XF86DGAGetVideo(dis, screen, addr, width, bank, ram)
Display *dis;
int screen;
char **addr;
int *width, *bank, *ram;
{
   int offset, fd;
#ifdef __EMX__
   APIRET rc;
   ULONG action;
   HFILE hfd;
#endif

   XF86DGAGetVideoLL(dis, screen , &offset, width, bank, ram);

#ifndef Lynx
#if defined(ISC) && defined(HAS_SVR3_MMAP)
   if ((fd = open("/dev/mmap", O_RDWR)) < 0)
   {
        fprintf(stderr, "XF86DGAGetVideo: failed to open /dev/mmap (%s)\n",
                           strerror(errno));
#else
#ifdef __EMX__
   /* Dragon warning here! /dev/pmap$ is never closed, except on progam exit.
    * Consecutive calling of this routine will make PMAP$ driver run out
    * of memory handles. Some umap/close mechanism should be provided
    */

   rc = DosOpen("/dev/pmap$", &hfd, &action, 0, FILE_NORMAL, FILE_OPEN,
		OPEN_ACCESS_READWRITE | OPEN_SHARE_DENYNONE, (PEAOP2)NULL);
   if (rc != 0) {
	fprintf(stderr, 
		"XF86DGAGetVideo: failed to open /dev/pmap$ (rc=%d)\n",
		rc);
#else
   if ((fd = open(DEV_MEM, O_RDWR)) < 0)
   {
        fprintf(stderr, "XF86DGAGetVideo: failed to open %s (%s)\n",
                           DEV_MEM, strerror(errno));
#endif
#endif
        exit (-1);
   }
#endif

#if defined(ISC) && defined(HAS_SVR3_MMAP)
   XFree86mloc.vaddr=(char *) 0;
   XFree86mloc.physaddr=(char *)offset;
   XFree86mloc.length=*bank;
   XFree86mloc.ioflg=1;

#define DEBUG_MMAP(state)     fprintf(stderr,"\
        XF86DGAGetVideo: (%s) \n\
            vaddr=%x/%x physaddr=%x/%x length=%d/%d\n",\
        (state) ,\
        XFree86mloc.vaddr,*addr,\
        XFree86mloc.physaddr,offset,\
        XFree86mloc.length,*bank)

   DEBUG_MMAP("before MMAP");
   if ((*addr = (char *) ioctl( fd, MAP, &XFree86mloc)) != (char *)-1) {
     offset=(int)XFree86mloc.physaddr;
     *bank=(int)XFree86mloc.length;
     DEBUG_MMAP("after MMAP");
   } else {
     DEBUG_MMAP("after MMAP");
     fprintf(stderr, "XF86DGAGetVideo: failed to mmap /dev/mmap (%s)\n",
                           strerror(errno));
#else /* !ISC */
#ifdef Lynx
   *addr = (void *)smem_create("XF86DGA", (char *)offset, *bank, SM_READ|SM_WRITE);
   if (*addr == NULL) {
        fprintf(stderr, "XF86DGAGetVideo: smem_create() failed (%s)\n",
                           strerror(errno));
#else /* !Lynx */
#ifdef __EMX__
   {
	struct map_ioctl {
		union {
			ULONG phys;
			void* user;
		} a;
		ULONG size;
	} pmap,dmap;
	ULONG plen,dlen;
#define XFREE86_PMAP	0x76
#define PMAP_MAP	0x44

	pmap.a.phys = offset;
	pmap.size = *bank;
	rc = DosDevIOCtl(hfd, XFREE86_PMAP, PMAP_MAP,
			 (PULONG)&pmap, sizeof(pmap), &plen,
			 (PULONG)&dmap, sizeof(dmap), &dlen);
	if (rc==0) {
		*addr = dmap.a.user;
	}
   }
   if (rc != 0) {
        fprintf(stderr, 
		"XF86DGAGetVideo: failed to mmap /dev/pmap$ (rc=%d)\n",
                rc);
#else /* !__EMX__ */
#ifndef MAP_FILE
#define MAP_FILE 0
#endif
   /* This requires linux-0.99.pl10 or above */
   *addr = (void *)mmap(NULL, *bank, PROT_READ,
                        MAP_FILE | MAP_SHARED, fd, (off_t)offset);
#ifdef DEBUG
   fprintf(stderr, "XF86DGAGetVideo: physaddr: 0x%08x, size: %d\n",
	   (long)offset, *bank);
#endif
   if (*addr == (char *) -1) {
        fprintf(stderr, "XF86DGAGetVideo: failed to mmap %s (%s)\n",
                           DEV_MEM, strerror(errno));
#endif /* !__EMX__*/
#endif /* !Lynx */
#endif /* !ISC && !HAS_SVR3_MMAP */
        exit (-2);
   }
   _XFree86size = *bank;
   _XFree86addr = *addr;

   
   atexit((void(*)(void))XF86cleanup);
   /* one shot XF86cleanup attempts */
   signal(SIGSEGV, XF86cleanup);
#ifdef SIGBUS
   signal(SIGBUS, XF86cleanup);
#endif
   signal(SIGHUP, XF86cleanup);
   signal(SIGFPE, XF86cleanup);  

   return 1;
}
